lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -322,6 +322,108 @@ void SPIRVRegularizeLLVMBase::expandSYCLTypeUsing(Module *M) {
     expandVIDWithSYCLTypeByValComp(F);
 }
 
+// In this function, we handle two conversion operations
+// 1. fptoui.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// 2. fptosi.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// Such non-standard integer types cannot be handled in SPIR-V. Hence, they
+// will be promoted to
+// 1. fptoui.sat.i64.fY (Y is 32 or 64)
+// 2. fptosi.sat.i64.fY (Y is 32 or 64)
+// However, LLVM documentation requires the following rules to be obeyed.
+// Rule 1: If the argument is any NaN, zero is returned.
+// Rule 2: If the argument is smaller than the smallest representable
+// (un)signed integer of the result type, the smallest representable
+// (un)signed integer is returned.
+// Rule 3: If the argument is larger than the largest representable (un)signed
+// integer of the result type, the largest representable (un)signed integer is
+// returned.
+// Rule 4: Otherwise, the result of rounding the argument towards zero is
+// returned.
+// Rules 1 & 4 are preserved when promoting iX to i64. For preserving Rule 2
+// and Rule 3, we saturate the result of the promoted instruction based on
+// original integer type (iX)
+// Example:
+// Input:
+// %0 = call i2 @llvm.fptoui.sat.i2.f32(float %input)
+// %1 = zext i32 %0
+// Output:
+// %0 = call i32 @_Z17convert_ulong_satf(float %input)
+// %1 = icmp uge i32 %0, 3 <Largest 2-bit unsigned integer>
+// %2 = icmp ule i32 %0, 0 <Smallest 2-bit unsigned integer>
```

Comment:
For conversion to an unsigned type, the output of the call will already guarantee a minimum value of 0, so there is no need for guard code on that side.

Reply:
That's a good point. Thanks. Is it ok to leave the code as is for sake of readability?

Reply:
Thinking again, seems like a useful change. Thanks. Addressed in the next commit.